### ISSUE TITLE
False positive [invalid-assignment] on assigning to union with bounded typevar

### ISSUE THREAD

**Issue by randolf-scholz** [astral-sh/ty#2314]
### Summary

https://play.ty.dev/b3098a7e-d5ce-480f-be95-3f4cd7a6d224

```python
def upcast(arg: int) -> int:
    return arg

def test[X: int](items: list[X]) -> list[X]:
    _a: list[int] = list(items)           # ✅️
    _b: list[int] = [*items]              # ✅️
    _c: list[int] = [x for x in items]    # ✅️

    _1: list[str] | list[int] = list(items)         # ❌️
    _2: list[int] | list[str] = list(items)         # ❌️
    _3: list[str] | list[int] = [*items]            # ✅️
    _4: list[int] | list[str] = [*items]            # ✅️
    _5: list[str] | list[int] = [x for x in items]  # ✅️
    _6: list[int] | list[str] = [x for x in items]  # ✅️
    _7: list[str] | list[int] = [upcast(x) for x in items]  # ✅️
    _8: list[int] | list[str] = [upcast(x) for x in items]  # ✅️
    return items
```

This reports

> Object of type `list[X@test]` is not assignable to `list[str] | list[int]

But at the same time, it reports `list` is `[T] (Iterable[T]) -> list[T]`.

So in `_1` and `_2` we have joint constraints:

- `X <: T` (`list[X]` must be assignable to `Iterable[T]`), 
- `X <: int` upper bound
- `T = int` or `T=str` (`list[T] <: list[int] | list[str]`)

So, this has the unique solution `T=int`, but `ty` seems to incorrectly select `T=X`.

**Comment by ElVynz** [comment]
I have encountered the same issue in a very similar case (https://play.ty.dev/77b3c10e-0aeb-4f2f-8007-31b79671733e)

**Comment by carljm** [comment]
Thanks for the report!

**Comment by Hugo-Polloli** [comment]
If that is alright, I would like to try to fix this one :)

**Comment by carljm** [comment]
> If that is alright, I would like to try to fix this one :)

Definitely!

**Comment by DetachHead** [comment]
another example:

```py
type MaybeList[T] = list[T] | T

def test[X: int](items: list[X]) -> list[X]:
    a: MaybeList[str | int] = list(items) # ❌️
    return items
```

https://play.ty.dev/1757f2f8-b251-4a94-96d2-51ff3743f6d9

**Comment by carljm** [comment]
@DetachHead I suspect that is something different from the OP case. It seems to be [fine using a PEP 613 type alias](https://play.ty.dev/99e56de2-304f-459e-9fcf-45f96597a927) instead of a PEP 695 one, so it's likely a place where we aren't handling PEP 695 aliases correctly.

**Comment by Hugo-Polloli** [comment]
This was fixed by https://github.com/astral-sh/ruff/pull/22124

The related PR https://github.com/astral-sh/ruff/pull/22426 only exists to (potentially) add a non-regression mdtest, but the false positive is effectively gone
